File: goldencheck/engine/validator.py
"""Validator — checks data against pinned rules in goldencheck.yml."""
from __future__ import annotations
import logging
from pathlib import Path
import polars as pl
from goldencheck.engine.reader import read_file
from goldencheck.config.schema import GoldenCheckConfig, ColumnRule
from goldencheck.models.finding import Finding, Severity

logger = logging.getLogger(__name__)


def validate_file(path: Path, config: GoldenCheckConfig) -> list[Finding]:
    df = read_file(path)
    findings: list[Finding] = []
    for col_name, rule in config.columns.items():
        if col_name not in df.columns:
            findings.append(Finding(
                severity=Severity.WARNING,
                column=col_name,
                check="existence",
                message=f"Column '{col_name}' defined in rules but not found in data",
            ))
            continue
        col = df[col_name]
        findings.extend(_check_column(col, col_name, rule))
    # Filter out ignored findings
    ignored = {(i.column, i.check) for i in config.ignore}
    findings = [f for f in findings if (f.column, f.check) not in ignored]
    findings.sort(key=lambda f: f.severity, reverse=True)
    return findings


def _check_column(col: pl.Series, name: str, rule: ColumnRule) -> list[Finding]:
    findings: list[Finding] = []

    # Required check
    if rule.required and col.null_count() > 0:
        findings.append(Finding(
            severity=Severity.ERROR,
            column=name,
            check="required",
            message=f"Required column has {col.null_count()} null values",
            affected_rows=col.null_count(),
        ))

    # Unique check
    if rule.unique:
        non_null = col.drop_nulls()
        dups = len(non_null) - non_null.n_unique()
        if dups > 0:
            findings.append(Finding(
                severity=Severity.ERROR,
                column=name,
                check="unique",
                message=f"Column should be unique but has {dups} duplicates",
                affected_rows=dups,
            ))

    # Enum check
    if rule.enum:
        non_null = col.drop_nulls().cast(pl.Utf8)
        invalid = non_null.filter(~non_null.is_in(rule.enum))
        if len(invalid) > 0:
            samples = invalid.head(5).to_list()
            findings.append(Finding(
                severity=Severity.ERROR,
                column=name,
                check="enum",
                message=f"{len(invalid)} values not in allowed enum {rule.enum}",
                affected_rows=len(invalid),
                sample_values=[str(s) for s in samples],
            ))

    # Range check
    if rule.range and len(rule.range) == 2:
        try:
            numeric = col.drop_nulls().cast(pl.Float64, strict=False).drop_nulls()
            lo, hi = rule.range
            out_of_range = numeric.filter((numeric < lo) | (numeric > hi))
            if len(out_of_range) > 0:
                samples = out_of_range.head(5).to_list()
                findings.append(Finding(
                    severity=Severity.ERROR,
                    column=name,
                    check="range",
                    message=f"{len(out_of_range)} values outside range [{lo}, {hi}]",
                    affected_rows=len(out_of_range),
                    sample_values=[str(s) for s in samples],
                ))
        except Exception:
            pass

    return findings

File: packages/goldencheck-js/src/core/engine/validator.ts
/**
 * Validator — checks TabularData against pinned rules in GoldenCheckConfig.
 * Port of goldencheck/engine/validator.py.
 */

import type { TabularData } from "../data.js";
import {
  type Finding,
  type GoldenCheckConfig,
  type ColumnRule,
  Severity,
  makeFinding,
} from "../types.js";
import { isNullish } from "../data.js";

/**
 * Validate data against a GoldenCheckConfig.
 * Checks: existence, required, unique, enum, range.
 * Filters out findings that match config.ignore entries.
 * Returns findings sorted by severity descending (ERROR first).
 */
export function validateData(
  data: TabularData,
  config: GoldenCheckConfig,
): Finding[] {
  const findings: Finding[] = [];
  const dataColumns = new Set(data.columns);

  for (const [colName, rule] of Object.entries(config.columns)) {
    if (!dataColumns.has(colName)) {
      findings.push(
        makeFinding({
          severity: Severity.WARNING,
          column: colName,
          check: "existence",
          message: `Column '${colName}' defined in rules but not found in data`,
        }),
      );
      continue;
    }

    findings.push(...checkColumn(data, colName, rule));
  }

  // Filter out ignored findings
  const ignored = new Set<string>(
    config.ignore.map((i) => `${i.column}\0${i.check}`),
  );
  const filtered = findings.filter(
    (f) => !ignored.has(`${f.column}\0${f.check}`),
  );

  // Sort by severity descending (ERROR=3 first)
  filtered.sort((a, b) => b.severity - a.severity);
  return filtered;
}

// ---------------------------------------------------------------------------
// Per-column checks
// ---------------------------------------------------------------------------

function checkColumn(
  data: TabularData,
  name: string,
  rule: ColumnRule,
): Finding[] {
  const findings: Finding[] = [];

  // Required check
  if (rule.required) {
    const nullCount = data.nullCount(name);
    if (nullCount > 0) {
      findings.push(
        makeFinding({
          severity: Severity.ERROR,
          column: name,
          check: "required",
          message: `Required column has ${nullCount} null values`,
          affectedRows: nullCount,
        }),
      );
    }
  }

  // Unique check
  if (rule.unique) {
    const nonNullValues = data.dropNulls(name);
    const uniqueSet = new Set(nonNullValues);
    const dups = nonNullValues.length - uniqueSet.size;
    if (dups > 0) {
      findings.push(
        makeFinding({
          severity: Severity.ERROR,
          column: name,
          check: "unique",
          message: `Column should be unique but has ${dups} duplicates`,
          affectedRows: dups,
        }),
      );
    }
  }

  // Enum check
  if (rule.enum && rule.enum.length > 0) {
    const allowedSet = new Set(rule.enum);
    const invalid: string[] = [];
    for (const v of data.column(name)) {
      if (isNullish(v)) continue;
      const s = String(v);
      if (!allowedSet.has(s)) {
        invalid.push(s);
      }
    }
    if (invalid.length > 0) {
      const samples = invalid.slice(0, 5);
      findings.push(
        makeFinding({
          severity: Severity.ERROR,
          column: name,
          check: "enum",
          message: `${invalid.length} values not in allowed enum [${rule.enum.join(", ")}]`,
          affectedRows: invalid.length,
          sampleValues: samples,
        }),
      );
    }
  }

  // Range check
  if (rule.range && rule.range.length === 2) {
    const [lo, hi] = rule.range;
    try {
      const outOfRange: string[] = [];
      for (const v of data.column(name)) {
        if (isNullish(v)) continue;
        const n = typeof v === "number" ? v : Number(v);
        if (!Number.isFinite(n)) continue;
        if (n < lo || n > hi) {
          outOfRange.push(String(v));
        }
      }
      if (outOfRange.length > 0) {
        const samples = outOfRange.slice(0, 5);
        findings.push(
          makeFinding({
            severity: Severity.ERROR,
            column: name,
            check: "range",
            message: `${outOfRange.length} values outside range [${lo}, ${hi}]`,
            affectedRows: outOfRange.length,
            sampleValues: samples,
          }),
        );
      }
    } catch {
      // Non-numeric column — skip range check silently
    }
  }

  return findings;
}
